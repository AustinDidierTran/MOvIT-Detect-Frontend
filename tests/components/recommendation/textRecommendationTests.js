/**
 * @author Gabriel Boucher
 * @author Anne-Marie Desloges
 * @author Austin-Didier Tran
 * @author Benjamin Roy
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'primereact/components/checkbox/Checkbox';
import { InputText } from 'primereact/components/inputtext/InputText';
import configureMockStore from 'redux-mock-store';
import sinon from 'sinon';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import TextRecommendation from '../../../src/components/recommendation/textRecommendation';

Enzyme.configure({ adapter: new Adapter() });

describe('TextRecommandation Tests', () => {
  const onChangeActiveSpy = sinon.spy();
  const onChangeValueSpy = sinon.spy();
  const initialState = { applicationReducer: { language: 'en' } };
  const mockStore = configureMockStore();
  const store = mockStore(initialState);
  const props = {
    recActive: true,
    title: 'Test',
    maxAngle: 45,
    value: 0,
    language: 'en',
    onChangeActive: (value) => { onChangeActiveSpy(value); },
    onChangeValue: (value) => { onChangeValueSpy(value); },
  };

  beforeEach(() => {
    onChangeValueSpy.resetHistory();
    onChangeActiveSpy.resetHistory();
  });

  it('should have proptypes', () => {
    const actualValue = TextRecommendation.WrappedComponent.propTypes;

    const expectedValue = {
      language: PropTypes.string.isRequired,
      recActive: PropTypes.bool.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string,
      onChangeActive: PropTypes.func.isRequired,
      onChangeValue: PropTypes.func.isRequired,
    };

    expect(JSON.stringify(actualValue)).toEqual(JSON.stringify(expectedValue));
  });

  it('should trigger onChangeActive when simulating a change event on checkbox', () => {
    const wrapper = shallow(<TextRecommendation store={store} {...props} />).dive();

    wrapper.find(Checkbox).simulate('change', { checked: false });

    expect(onChangeActiveSpy.calledOnce).toEqual(true);
    expect(onChangeActiveSpy.getCalls()[0].args[0]).toEqual(false);
  });

  it('should trigger onChangeValue with an empty string when simulating a change event on checkbox', () => {
    const wrapper = shallow(<TextRecommendation store={store} {...props} />).dive();

    wrapper.find(Checkbox).simulate('change', { checked: false });

    expect(onChangeValueSpy.calledOnce).toEqual(true);
    expect(onChangeValueSpy.getCalls()[0].args[0]).toEqual('');
  });

  it('should trigger onChangeValue with a parameter when simulating a change event on checkbox', () => {
    const wrapper = shallow(<TextRecommendation store={store} {...props} />).dive();

    wrapper.find(Checkbox).simulate('change', { checked: true });

    expect(onChangeValueSpy.calledOnce).toEqual(true);
    expect(onChangeValueSpy.getCalls()[0].args[0]).toEqual('recommendations.tiltAsNeeded.en');
  });

  it('should trigger onChangeValue when simulating a change event on InputText', () => {
    const wrapper = shallow(<TextRecommendation store={store} {...props} />).dive();

    wrapper.find(InputText).simulate('change', { target: { value: 'Test' } });

    expect(onChangeValueSpy.calledOnce).toEqual(true);
    expect(onChangeValueSpy.getCalls()[0].args[0]).toEqual('Test');
  });

  it('should match the snapshot when the recommendation is active', () => {
    const wrapper = shallow(<TextRecommendation store={store} {...props} />).dive();

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match the snapshot when the recommendation is inactive', () => {
    const inactiveProps = {
      recActive: false,
      title: 'Test',
      maxAngle: 45,
      value: 10,
      language: 'en',
    };
    const wrapper = shallow(<TextRecommendation store={store} {...inactiveProps} />).dive();

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
