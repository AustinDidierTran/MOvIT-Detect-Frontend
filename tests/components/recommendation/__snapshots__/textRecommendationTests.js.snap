// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextRecommandation Tests should match the snapshot when the recommendation is active 1`] = `
<div
  className="pt-2 pl-3"
>
  <Checkbox
    checked={true}
    className={null}
    disabled={false}
    id={null}
    inputId="activeRecCheck"
    label="Test"
    name={null}
    onChange={[Function]}
    onContextMenu={null}
    onMouseDown={null}
    readOnly={false}
    style={null}
    value={null}
  />
  <label
    className="mt-1"
    htmlFor="activeRecCheck"
  >
    Test
  </label>
  <div
    className="col-12 col-md-9 mb-2 ml-3"
  >
    <InputText
      className="form-control"
      id="textRec"
      keyfilter={null}
      onChange={[Function]}
      onInput={null}
      onKeyPress={null}
      placeholder="recommendations.tiltAsNeeded.en"
      type="text"
      validateOnly={false}
      value=""
    />
  </div>
  <Tooltip
    appendTo="body"
    escape={true}
    for="#textRecInfoundefined"
    hideDelay={null}
    onBeforeShow={null}
    positionStyle={null}
    showDelay={null}
    title={null}
    tooltipDisabled={false}
    tooltipEvent="hover"
    tooltipPosition="right"
    tooltipStyleClass={null}
  />
</div>
`;

exports[`TextRecommandation Tests should match the snapshot when the recommendation is inactive 1`] = `
<div
  className="pt-2 pl-3"
>
  <Checkbox
    checked={false}
    className={null}
    disabled={false}
    id={null}
    inputId="activeRecCheck"
    label="Test"
    name={null}
    onChange={[Function]}
    onContextMenu={null}
    onMouseDown={null}
    readOnly={false}
    style={null}
    value={null}
  />
  <label
    className="mt-1"
    htmlFor="activeRecCheck"
  >
    Test
  </label>
  <Tooltip
    appendTo="body"
    escape={true}
    for="#textRecInfoundefined"
    hideDelay={null}
    onBeforeShow={null}
    positionStyle={null}
    showDelay={null}
    title={null}
    tooltipDisabled={false}
    tooltipEvent="hover"
    tooltipPosition="right"
    tooltipStyleClass={null}
  />
</div>
`;
